<a href="https://colab.research.google.com/github/MohdHamza2/Health-assistent/blob/main/Task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze the restaurant dataset from "/content/Dataset .csv" to build a machine learning model that predicts the aggregate rating based on other features. The analysis should include data preprocessing, model training and evaluation, and interpretation of the results to identify influential features.

## Load the data

### Subtask:
Load the dataset from `/content/Dataset .csv` into a pandas DataFrame.


**Reasoning**:
Load the dataset into a pandas DataFrame and display its head and shape.



In [1]:
import pandas as pd

df = pd.read_csv('/content/Dataset .csv')
display(df.head())
print(df.shape)

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

(9551, 21)


## Handle missing values

### Subtask:
Inspect the data for missing values and handle them appropriately (e.g., imputation or removal).


**Reasoning**:
Check for missing values in each column and decide how to handle them.



In [2]:
print("Missing values before handling:")
print(df.isnull().sum())

# Determine strategy: If a column has a small number of missing values, impute or drop rows.
# If a column has a large number of missing values, consider dropping the column.
# In this dataset, 'Cuisines' is the only column with missing values and only 9 rows are missing.
# We will drop these rows.
df.dropna(subset=['Cuisines'], inplace=True)

print("\nMissing values after handling:")
print(df.isnull().sum())

Missing values before handling:
Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

Missing values after handling:
Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering

## Encode categorical variables

### Subtask:
Convert categorical features into a numerical format suitable for machine learning models.


**Reasoning**:
Identify categorical columns and apply one-hot encoding.



In [3]:
# Identify categorical columns
categorical_cols = df.select_dtypes(include='object').columns

# Apply one-hot encoding
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

# Display the first few rows of the modified DataFrame
display(df.head())

Restaurant ID  Country Code   Longitude   Latitude  Average Cost for two  \
0        6317637           162  121.027535  14.565443                  1100   
1        6304287           162  121.014101  14.553708                  1200   
2        6300002           162  121.056831  14.581404                  4000   
3        6318506           162  121.056475  14.585318                  1500   
4        6314302           162  121.057508  14.584450                  1500   

   Price range  Aggregate rating  Votes  Restaurant Name_#Dilliwaala6  \
0            3               4.8    314                         False   
1            3               4.5    591                         False   
2            4               4.4    270                         False   
3            4               4.9    365                         False   
4            4               4.8    229                         False   

   Restaurant Name_#InstaFreeze  ...  Rating color_Green  Rating color_Orange  \
0                         False  ...               False                False   
1                         False  ...               False                False   
2                         False  ...                True                False   
3                         False  ...               False                False   
4                         False  ...               False                False   

   Rating color_Red  Rating color_White  Rating color_Yellow  \
0             False               False                False   
1             False               False                False   
2             False               False                False   
3             False               False                False   
4             False               False                False   

   Rating text_Excellent  Rating text_Good  Rating text_Not rated  \
0                   True             False                  False   
1                   True             False                  False   
2                  False             False                  False   
3                   True             False                  False   
4                   True             False                  False   

   Rating text_Poor  Rating text_Very Good  
0             False                  False  
1             False                  False  
2             False                   True  
3             False                  False  
4             False                  False  

[5 rows x 20807 columns]

## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the data into features (X) and the target variable (y), then split these into training and testing sets.



In [4]:
from sklearn.model_selection import train_test_split

X = df.drop('Aggregate rating', axis=1)
y = df['Aggregate rating']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (7633, 20806)
Shape of X_test: (1909, 20806)
Shape of y_train: (7633,)
Shape of y_test: (1909,)


## Select and train a model

### Subtask:
Choose a regression algorithm and train it on the training data.


**Reasoning**:
Train a Random Forest Regressor model on the training data.



In [5]:
from sklearn.ensemble import RandomForestRegressor

# Instantiate the model
model = RandomForestRegressor(random_state=42)

# Train the model
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## Evaluate the model

### Subtask:
Evaluate the trained model's performance using appropriate regression metrics on the testing data.


**Reasoning**:
Use the trained model to make predictions on the test set and then calculate and print the evaluation metrics as instructed.



In [6]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate evaluation metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

Mean Absolute Error (MAE): 0.11107857517024625
Mean Squared Error (MSE): 0.02899933525405971
R-squared (R2) Score: 0.987336712890049


## Interpret the model

### Subtask:
Analyze the model's results to identify the most influential features affecting restaurant ratings.


**Reasoning**:
Access the feature importances from the trained model and create a DataFrame to store the feature names and their corresponding importances. Then sort the feature importances and print the top 10 most influential features.



In [7]:
# Get feature importances
feature_importances = model.feature_importances_

# Create a pandas Series of feature importances
feature_importances_series = pd.Series(feature_importances, index=X_train.columns)

# Sort feature importances in descending order
sorted_feature_importances = feature_importances_series.sort_values(ascending=False)

# Print the top 10 most influential features
print("Top 10 Most Influential Features:")
print(sorted_feature_importances.head(10))

Top 10 Most Influential Features:
Rating color_White       0.367875
Votes                    0.280427
Rating text_Not rated    0.251132
Rating color_Orange      0.052230
Rating text_Poor         0.011944
Rating color_Red         0.008613
Rating color_Yellow      0.008537
Rating text_Good         0.004999
Restaurant ID            0.002077
Longitude                0.001149
dtype: float64


## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded and contained 9551 rows and 21 columns.
*   Missing values were only present in the 'Cuisines' column (9 rows), which were subsequently dropped.
*   Categorical features were successfully one-hot encoded, increasing the number of columns in the dataset.
*   The dataset was split into training (80%) and testing (20%) sets, resulting in 7633 training samples and 1909 testing samples.
*   A Random Forest Regressor model was trained on the training data.
*   The trained model achieved the following performance metrics on the test set:
    *   Mean Absolute Error (MAE): 0.1111
    *   Mean Squared Error (MSE): 0.0290
    *   R-squared (R2) Score: 0.9873
*   The top 3 most influential features identified by the model were 'Rating color\_White', 'Votes', and 'Rating text\_Not rated'.

### Insights or Next Steps

*   The Random Forest model demonstrated excellent performance in predicting aggregate ratings based on the provided features.
*   Further investigation into the relationship between 'Rating color\_White', 'Votes', 'Rating text\_Not rated' and the aggregate rating could provide deeper insights into the factors driving restaurant ratings.


## Summary:

### Data Analysis Key Findings

* The dataset was successfully loaded and contained 9551 rows and 21 columns.
* Missing values were only present in the 'Cuisines' column (9 rows), which were subsequently dropped.
* Categorical features were successfully one-hot encoded, increasing the number of columns in the dataset.
* The dataset was split into training (80%) and testing (20%) sets, resulting in 7633 training samples and 1909 testing samples.
* A Random Forest Regressor model was trained on the training data.
* The trained model achieved the following performance metrics on the test set:
  * Mean Absolute Error (MAE): {{mae:.4f}}
  * Mean Squared Error (MSE): {{mse:.4f}}
  * R-squared (R2) Score: {{r2:.4f}}
* The top 3 most influential features identified by the model were 'Rating color\_White', 'Votes', and 'Rating text\_Not rated'.

### Insights or Next Steps

* The Random Forest model demonstrated excellent performance in predicting aggregate ratings based on the provided features.
* Further investigation into the relationship between 'Rating color\_White', 'Votes', 'Rating text\_Not rated' and the aggregate rating could provide deeper insights into the factors driving restaurant ratings.

In [8]:
import joblib

# Save the trained model to a file
filename = 'random_forest_restaurant_rating_model.joblib'
joblib.dump(model, filename)

print(f"Model saved as {filename}")

Model saved as random_forest_restaurant_rating_model.joblib


In [9]:
import joblib

# Load the model from the file
loaded_model = joblib.load('random_forest_restaurant_rating_model.joblib')

# Now you can use the loaded_model for predictions or further analysis
# For example, to make a prediction on new data (assuming new_data is a pandas DataFrame):
# prediction = loaded_model.predict(new_data)